src/scmdata/groupby.py

```diff
@@ -114,9 +123,67 @@ def apply(self, func, *args, **kwargs):
             The result of splitting, applying and combining this array.
         """
         grouped = self._iter_grouped()
-        applied = [
-            _maybe_wrap_array(arr, func(arr, *args, **kwargs)) for arr in grouped
-        ]
+        applied = [func(arr, *args, **kwargs) for arr in grouped]
+        return self._combine(applied)
+
+    def apply_parallel(
+        self,
+        func: "Callable[Concatenate[GenericRun, P], Union[GenericRun, pd.DataFrame, None]]",
+        n_jobs: int = 1,
+        backend: str = "loky",
+        *args: "P.args",
+        **kwargs: "P.kwargs",
+    ) -> "GenericRun":
+        """
+        Apply a function to each group in parallel and append the results
+
+        Provides the same functionality as :func:`~apply` except that :mod:`joblib` is used to apply
+        `func` to each group in parallel. This can be slower than using :func:`~apply` for small
+        numbers of groups or in the case where `func` is fast as there is overhead setting up the
+        processing pool.
+
+        See Also
+        --------
+        :func:`~apply`
+
+        Parameters
+        ----------
+        func
+            Callable to apply to each timeseries.
+
+        n_jobs
+            Number of jobs to run in parallel (defaults to a single job which is useful for
+            debugging purposes). If `-1` all CPUs are used.
+
+        backend
+            Backend used for parallelisation. Defaults to 'loky' which uses separate processes for
+            each worker.
+
+            See :class:`joblib.Parallel` for a more complete description of the available
+            options.
+
+        ``*args``
+            Positional arguments passed to `func`.
+
+        ``**kwargs``
+            Used to call `func(ar, **kwargs)` for each array `ar`.
+
+        Returns
+        -------
+        applied : :class:`ScmRun <scmdata.run.ScmRun>`
+            The result of splitting, applying and combining this array.
+        """
+        try:
+            import joblib  # type: ignore
```

Comment:
Is joblb preferred for this sort of stuff over using e.g. the multiprocessing module? It probably doesn't matter much, but would be good to understand the reasoning a bit more here so we can decide whether we want to use dependency injection to give the user more control if they want. (cc @mikapfl)

Reply:
Ok. I'm happy to discuss in follow up MRs rather than here, but for reference my thoughts are:

The way this is currently written, to test it, we would often have to mock out joblib. That suggests this function is doing too much to me (it is both setting up the joblib pool and executing it). Re-writing so that the pool can be created elsewhere (with a helper for users that creates the pool if it isn't supplied) and then this method just uses the pool would separate those two things and would make it possible to test without having to mock joblib.

I started having a go at this and I think breaking it up is pretty simple (see below). The thing that exploded on me was type hints (hence why it might be much simpler to discuss in follow up, self-contained MRs).

```python
# break up of method
def apply_parallel(
    self,
    func,
    *args,
    pool = None,
    group_preparer = None,
    **kwargs,
):
    if pool is None:
        pool = get_joblib_pool()

    if group_preparer is None:
        group_preparer = get_joblib_group_preparer()

    grouped = self._iter_grouped()

    applied = pool(group_preparer(func, group, *args, **kwargs) for group in grouped)
    return self._combine(applied)
```

Reply:
Same idea as how `get_ymax` is handled here https://github.com/openscm/OpenSCM-Calibration/blob/1f3d798730b13338f699310dc1db4e96b12aea9e/src/openscm_calibration/scipy_plotting.py#L507

Reply:
Implementation is in https://github.com/openscm/scmdata/pull/264/files

Happy to either merge that in here or do it as part of a follow up (follow up may be easier as ruff got hold of the commits and changed more than I was hoping).

Reply:
Zeb's follow up MR came out much cleaner than I thought so I do think it is an improvement even if it doesn't add any additional tests. Not sure if we will ever need to inject different functionality, but it's a small price